### DISCUSSION TITLE
Filling in prompt fields based on other prompts dynamically

### DISCUSSION THREAD

**Discussion by astro-stan** [goauthentik/authentik#2107] [Ideas]
Let's say I have the following registration prompt:
![image](https://user-images.githubusercontent.com/36302090/150051255-5778b50e-b759-4884-9a27-2ebb7529138e.png)

It would have been great if we could use an expression policy (or something else) to fill in the `username` field automatically with a suggested value based on input in the `first_name` and `last_name` prompts while the user is still filling in the form.

Example:
1. User types 'John' in the `first_name` prompt
2. `username` prompt fills in with the value of: `john`
3. User continues to type 'Doe' in the `last_name` prompt
4. `username` prompt fills up with the value of: `john.doe`
5. User proceeds to fill in the rest of the prompts and submits the form
6. A new user is created with the username of `john.doe`

**Comment by BeryJu** [comment]
Agree that this would be a nice feature, sadly not currently possible since policies are only run upon stage completion, you could break the flow up into two separate prompt stage and use the policies inbetween to generate data.

For this we'd have to either allow admins to run custom javascript in the flow executor, or send the data to the backend on every keystroke which is also not ideal.

**Comment by astro-stan** [comment]
Thanks for replying.

Yes, I eventually split it into multiple stages. However, I was unable to pre-populate the field with an expression policy. I only managed to replace its placeholder.

Here are the two expression policies I used:
1. This one sanitizes the first and last names the new user inputs in the first stage of the registration process:
```
# Sanitizes the first and last name attributes by striping whitespace from both ends
# and matching against a regex to ensure only valid names are entered

# ak_logger.debug(request.context)

first_name = request.context['prompt_data']['first_name'].strip()
last_name = request.context['prompt_data']['last_name'].strip()

name_regex = '^[a-zA-Z]{2,}$'
err_msg = (
  'must consist of only small or capital leters,'
  'be at least 2 symbols long and contain no spaces.'
)
if not regex_match(first_name, name_regex):
  ak_message('First name ' + err_msg)
  return False
if not regex_match(last_name, name_regex):
  ak_message('Last name ' + err_msg)
  return False

request.context['prompt_data']['last_name'] = last_name
request.context['prompt_data']['first_name'] = first_name

return True
```

2. This one runs after the first one and uses the first and last name to create a user name and a display name.
```
# Trust the 'first_name' and 'last_name' fields to be sanitized and valid
# Use them to create a full display name and username

first_name = request.context['prompt_data']['first_name']
last_name = request.context['prompt_data']['last_name']
request.context['prompt_data']['name'] = first_name + " " + last_name
request.context['prompt_data']['username'] = first_name + "." + last_name

return True
```

I do not have an input for the display name, so that is created directly by the expression policy, with no option for the user to change it (except when filling in his first and last name). However, I do have an "username" field, which I would like to pre-fill but still leave the user the flexibility to change it if they desire.

Unfortunately, the expression policy only fills in the username as a placeholder and the user is still required to type their username (screenshot & context below)

![image](https://user-images.githubusercontent.com/36302090/152034241-97f09141-fafa-48a9-badb-ef43028d730d.png)

The current plan context:
```
{
    "invitation_in_effect": true,
    "invitation": {
        "app": "authentik_stages_invitation",
        "model_name": "invitation",
        "pk": "8e2e...",
        "name": "Invitation 8e2e..."
    },
    "prompt_data": {
        "component": "ak-stage-prompt",
        "first_name": "my",
        "last_name": "name",
        "name": "my name",
        "username": "my.name"
    }
}
```

Is this a bug or expected behaviour? Is it possible to actually pre-fill the username field?

**Comment by sevmonster** [comment]
@sdimovv Does https://github.com/goauthentik/authentik/pull/2397 help your use case?

**Reply by astro-stan** [reply]
> @sdimovv Does #2397 help your use case?

No, I don't think so.

What I'm trying to achieve is to pre-polulate fields dynamically based on other fields which the user has already filled in.

For example:
1. User enters first name "John"
2. Username field pre-polulates with "john"
3. User enters last name of "Doe"
4. Username field pre-polulates with "john.doe"
5.  User can now either skip filling in the username field and use the pre-populated value or delete it and enter a different username

**Reply by sevmonster** [reply]
Only reason I mentioned #2397 was because one of the features added to support the main feature was allowing expressions in placeholders, which I thought could help with your policy writing. You still won't get your fields updated at runtime, however.
